### ISSUE TITLE
RFC: Should matmuls use tf32 by default?

### ISSUE THREAD

**Issue by ngimel** [pytorch/pytorch#67384]
Since the release of Ampere GPUs, pytorch has been using [tf32](https://pytorch.org/docs/master/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices) by default.  It is providing much better performance at the expense of somewhat lower accuracy. Nvidia has conducted a lot of experiments proving that convergence behavior of a wide variety of networks does not change when tf32 is used instead of regular fp32. 
However, since pytorch is used not only for deep learning workloads, or for some non-standard deep learning workloads, the use of tf32 for matrix multiplication has resulted in [a lot](https://discuss.pytorch.org/t/bug-matmul-seems-to-cast-to-float16-internally/135248) of [confusion](https://github.com/pytorch/pytorch/issues?q=is%3Aissue+label%3A%22module%3A+tf32%22) and sometimes [bad results](https://github.com/pytorch/pytorch/issues/61890).
Going forward, we have a few options
1) Do nothing. Pros: most users get speedups, cons: issues and confusion will continue
2) Leave the default as is, but add warning to the first call using tf32, explicit setting of `allow_tf32` would silence this warning (proposed [here](https://github.com/pytorch/pytorch/issues/67185#issuecomment-951321420). Pros - results will no longer be a surprise, and users will be able to get the expected results without resorting to documentation. Cons - pytorch doesn't warn for normal operations, so this would create bad precedent
3) Use tf32 in nn layers (linear, rnn etc), but disallow it for raw matmul operations (proposed [here](https://github.com/pytorch/pytorch/issues/61890#issuecomment-906728150). Pros - networks using `nn` operations will get speed up while people experimenting with pure matmul operations will still get exact results. Cons - it is really confusing when linear layer that calls `matmul` under the hood produces different results than `matmul` call. Also, if someone defines a network using `matmuls` and not `nn.Linear` or `nn.linear` they won't see the speed-up.
4) Disable tf32 matmul by default, possibly with the warning that it can be enabled. Pros - no surprise, accurate results. Cons - users who are getting speed-ups currently will have to manually enable tf32, or see reduced performance. If there's a warning, same con as option 2). 
In my conversations with power users, they lean towards option 4), we should use this issue to find an acceptable solution. 
cc @zasdfgbnm @ptrblck @CrisHY1995, @ssnl, @stas00, @t-vi, @ptrblck, @csarofeen, @zasdfgbnm, @wjablonski-work, please copy anyone else who is interested in this discussion.

**Comment by Lezcano** [comment]
+1 for (4) without warning. I really like @stas00 's idea of having a "performance" document in the docs where we compile general tricks / guidance for how to do this, in a similar way as we do for [reproducibility](https://pytorch.org/docs/stable/notes/randomness.html).

This solution would also go in line with the `use_deterministic_algorithms` flag and others that are being discussed, e.g. https://github.com/pytorch/pytorch/pull/67980

**Comment by stas00** [comment]
Here is an update to my [initial comment](https://github.com/pytorch/pytorch/issues/67384#issuecomment-953416737).

We had a call with the NVIDIA engineers who kindly spent their time to show us a lot of benchmarks on trained models they have done with the TF32-enabled pytorch and that the outcomes were on par with the FP32 results quality-wise.

We are yet to have our own experience to validate that, since the Amphere hardware is still scarce and so it'll take more time before we get a first hand experience.

**Therefore it probably does make sense to have TF32 enabled by default**, if it's very easy for the user to find the documentation that explains the difference nuances of speed vs. correctness, how to turn different knobs on/off, what culprits to avoid, when not to use TF32, etc. 

And to make it easy to find cross-link to that document from more than one place (where it's relevant of course).

I imagine such document would have main sections like:

- How can I make my training/inference run faster
- How can I make my training/inference use less GPU memory
- How can I reduce the number of instabilities in my training

with performance-specific questions, addressing both hardware and software aspects.

And of course nice visuals that help the user to understand the difference:
https://blogs.nvidia.com/wp-content/uploads/2020/05/tf32-Mantissa-chart-hi-res-FINAL-400x255.png.webp
and probably a brief table of limitations - e.g. max number, min number of each type.

Thank you.

**Comment by ngimel** [comment]
@stas00 yes, tf32 shows great results with training nn models, and that's why it would make perfect sense to have it enabled by default in the Huggingface codebase. However, having it by default in pytorch has proven problematic, with many users facing accuracy issues. We have multiple docs mentioning tf32 and how to turn it off for more accurate results, and its associated tradeoffs, and yet that doesn't prevent github and forum issues.

**Comment by stas00** [comment]
Oh, OK, then I was under the wrong impression that my suggestion was impacting the pytorch choice. 

If as you're saying this is far more than just a preference for HF `transformers` then by all means please ignore my suggestion to what the ideal default might be.

Most likely we are going to choose the default for HF `transformers` once we get our hands on the hardware and get to run the quality tests ourselves.

**Comment by turtleizzy** [comment]
As a user and victim of tf32 I would strongly vote for (3) or (4) because torch gpu linear algebra operations are so commonly used in the training process to generate data (rotational & elastic transformation with `grid_sample`, etc)  and calculate metrics. 

In fact, issue #75778 was detected after several days of debugging for the performance deterioration on Ampere devices, which was eventually tracked down to low-quality input images after gpu transformation, just as #61890 illustrated. We have used ampere devices for quite a long time for training (with gpu image augmentation implemented by torch linalg ops) and we are now worried that the results we got on these devices were neither reliable nor reproducible (on other devices) due to silent inconsistency of input data that CNNs were notoriously sensitive about. Any hint from the module, as suggested by #67185, would have been useful.

**Comment by blefaudeux** [comment]
Adding two cents here following a discussion with @fmassa: one side effect of changing the default computation format is that unit tests using pytorch against custom (fp32) kernels suddenly fail when changing GPUs. It's fine once you know of this underlying change (ie: it was ok for us) but it can be a little disrupting. No biggie, just sharing that here

**Comment by mruberry** [comment]
Thanks everyone for their feedback! Based on the comments here we have now changed the default fp32 math mode in PyTorch. See https://github.com/pytorch/pytorch/pull/76509. There may be additional refinements we want to make to how users can control math modes and precision going forward, and we'll have different issues for those, but this particular question is resolved!
